File: ruby/lib/onvo/widgets.rb
# frozen_string_literal: true

require_relative '../resource'

# Dashboard Widget endpoints
class Widgets < Resource
  def initialize(endpoint, api_key, questions)
    super(endpoint, api_key)
    @questions = questions
  end

  # List out the widgets linked to a dashboard
  #
  # @param dashboard_id [String] The dashboard id to list widgets for
  # @return [Array] A list of all the widgets
  def list(dashboard_id)
    base_get('/widgets', query: { dashboard: dashboard_id })
  end

  # Get the details of a widget
  #
  # @param widget_id [String] The widget id to get
  # @return [Hash] The details of the widget
  def get(widget_id)
    base_get("/widgets/#{widget_id}")
  end

  # Create a new widget
  #
  # @param dashboard_id [String] The dashboard id to create the widget in
  # @param query [String] The query upon which a widget is created
  # @return [Hash] The details of the created widget
  def create(dashboard_id, query)
    conversation = @questions.create(dashboard_id, query)['messages']
    raw_widget_text = conversation.filter { |message| message['role'] == 'tool' }.last['content']

    code = raw_widget_text[/```python(?<code>[\S\n\t\v ]*)```\n```json/, :code].strip
    cache = JSON.parse(raw_widget_text[/```json(?<cache>[\S\n\t\v ]*)```/, :cache].strip)

    data = { dashboard: dashboard_id, code: code, cache: cache, messages: conversation }

    base_put('/widgets', body: data)
  end

  # Update the details of a widget. Note : updated data is not re-analyzed by the LLM
  #
  # @param widget_id [String] The widget id
  # @param body [Hash] The details to update
  # @return [Hash] The details of the updated widget
  def update(widget_id, body)
    base_post("/widgets/#{widget_id}", body: body)
  end

  # Delete a widget
  #
  # @param widget_id [String] The widget id to delete
  # @return [Hash] status of delete. Returns {"success": true} if successful.
  def delete(widget_id)
    base_delete("/widgets/#{widget_id}")
  end

  # Export a widget
  #
  # @param widget_id [String] The widget id to export
  # @param format [String] The export format
  # @return [Any] The widget in the requested format
  def export(widget_id, format)
    base_get("/widgets/#{widget_id}/export?format=#{format}")
  end

  # Request an edit to a widget. Note : updated data is re-analyzed by the LLM
  #
  # @param widget_id [String] The widget id to edit
  # @param body [Hash] The details of the edit
  # @return [Hash] The details of the updated widget
  def request_edit(widget_id, body)
    base_patch("/widgets/#{widget_id}", body: body)
  end

  # Execute code in the context of a widget. Note : The code uploaded does not modify the widget. This endpoint is used to ensure that the widget code works.
  #
  # @param widget_id [String] The widget id
  # @param code [String] The code to execute
  # @return [Hash] The details of the widget that would be created
  def execute_code(widget_id, code)
    base_post("/widgets/#{widget_id}/execute-code", body: { code: code })
  end
end

File: python/onvo/resources/widgets.py
from onvo.base_resource import BaseResource
import re
import json


class Widgets(BaseResource):
    def __init__(self, endpoint, api_key, questions):
        super().__init__(endpoint, api_key)
        self.questions = questions

    def list(self, dashboard_id):
        """List out the widgets linked to a dashboard

        Args:
            dashboard_id (str): The dashboard id to list widgets for

        Returns:
            list: A list of all the widgets
        """
        return self.base_get("/widgets", params={"dashboard": dashboard_id})

    def get(self, id):
        """Get details on a specific widget

        Args:
            id (str): The widget id to get details on

        Returns:
            dict: A dictionary containing details on the widget
        """
        return self.base_get(f"/widgets/{id}")

    def export(self, id, format):
        """Export a widget

        Args:
            id (str): The widget id to export
            format (str): The export format to export the widget in. Must be one of: png, jpg, svg

        Returns:
            Any: The file in the requested format
        """
        return self.base_get(f"/widgets/{id}/export?format={format}")

    def create(self, dashboard_id: str, query: str):
        """Create a new widget

        Args:
            dashboard_id (str): The dashboard id to create the widget for
            query (str): The query string to ask a question

        Returns:
            dict: A dictionary containing the details of the newly created widget
        """
        conversation = self.questions.create(dashboard_id, query)
        question_text = conversation["messages"][-1]["content"]

        def extract_pattern(pattern):
            return re.findall(pattern, question_text)[0].strip()

        code = extract_pattern(r"```python([\S\n\t\v ]*)```\n```json")
        cache = extract_pattern(r"```json([\S\n\t\v ]*)```")

        return self.base_put(
            "/widgets",
            data={
                "dashboard": dashboard_id,
                "code": code,
                "cache": json.loads(cache),
                "messages": [{"role": "user", "content": query}],
            },
        )

    def update(self, id, data):
        """Update an existing widget. Note: the code is not re-run when using this endpoint.

        Args:
            id (str): The widget id to update
            data (dict[str, Any]): The data within the widget to be updated

        Returns:
            dict: A dictionary containing the details of the updated widget
        """
        return self.base_post(f"/widgets/{id}", data=data)

    def request_edit(self, id, data):
        """Request an edit to an existing widget. Note: the code will be rebuilt based on changes to the messages

        Args:
            id (str): The widget id to request an edit
            data (dict[str, Any]): The data within the widget to be updated

        Returns:
            dict: A dictionary containing the details of the updated widget
        """
        return self.base_patch(f"/widgets/{id}", data=data)

    def execute_code(self, id, code):
        """Execute code. The code given to this endpoint will be run in the context of the widget. The code in the widget will not be replaced. Use this endpoint as a test bed to verify that your edits to the widget work.

        Args:
            id (str): The widget id to execute the code in
            code (str): The code to execute in the widget

        Returns:
            dict: A dictionary containing the details of the updated widget
        """
        return self.base_post(f"/widgets/{id}/execute-code", data={"code": code})

    def delete(self, id):
        """Delete an existing widget

        Args:
            id (str): The widget id to delete

        Returns:
            dict: A dictionary that shows status of the delete. Returns {"success": true} if successful.
        """
        return self.base_delete(f"/widgets/{id}")
